### PR TITLE
Proof of concept TCP server mode

### PR DESCRIPTION
**Pull request by tarruda**
This builds on my [other PR](https://github.com/ggerganov/llama.cpp/pull/267) to implement a very simple TCP mode.

The new mode first loads the model then listens for TCP connections on a port. When a connection is received, arguments will be parsed using a simple protocol:
    
- First the number of arguments will be read followed by a newline character.
- Then each argument will be read, separated by the 0 byte.
- With this we build an argument vector, similar to what is passed to  the program entry point. 
- The resulting "argv"  is passed `gpt_params_parse`.
    
Finally `llama_main` will be executed with the input/output streams connected to the socket.

I've included two sample bash scripts which can be used to test the new mode. This is how it works:

- Run `./chat_tcp_server.sh` in a terminal.
- In a second terminal, run `./chat_tcp_client.sh`. This will connect to the server and start a sample chat session.

One thing to note is that this mode is only implemented for Unixes. There's two reasons for that:

- I never wrote win32 TCP code, so I'm not familiar with the API
- There's really no advantage in implementing this for win32 because it doesn't support `fork()`. The main advantage of using this mode is that it serves each connection in a separate process which inherits memory from the parent (so the model only has to be loaded once).

While the protocol is a bit "low level", it should be easy to write a higher level API on top of this, such as a node.js web server or next.js app.

### CODE REVIEW

**Comment by ggerganov** [comment]
Shall we close #267 now that we have this?

**Comment by tarruda** [comment]
> Shall we close #267 now that we have this?

We can, this already includes all changes in #267. Only reason I did in a separate PR was to simplify reviewing.

**Comment by tarruda** [comment]
Rebased

**Comment by avilum** [comment]
If anyone is seeking for working client/server implementation,
I wrote a minimal Go/Python server and client with live streaming, that is based on this awesome repo.
See https://github.com/avilum/llama-saas

File: CMakeLists.txt
Comment (maxsu) [inline CMakeLists.txt:1]: We don't need to introduce the LLAMA_SRC variable. Instead we add the tcp_server.cpp source to the llama target using the target_sources function:

add_executable(llama
    llama.cpp
    main.cpp
    utils.cpp
    utils.h)

if(NOT WIN32)
    target_sources(llama PRIVATE tcp_server.cpp)
endif()

See: 
sample commit : https://github.com/maxsu/llama.cpp/commit/f5201905ff75cb7c13e1d485939e0ea159a9e096?diff=split

CI results: https://github.com/maxsu/llama.cpp/actions/runs/4462657355

Let me know if I can help you tidy up any other cmake code! I'm on a cmake binge!

Comment (tarruda) [reply CMakeLists.txt:1]: Awesome, I will adapt. Thanks

Comment (tarruda) [reply CMakeLists.txt:1]: Amended the last commit with your suggestion.

**Comment by tarruda** [comment]
> If anyone is seeking for working client/server implementation, I wrote a minimal Go/Python server and client with live streaming, that is based on this awesome repo. See https://github.com/avilum/llama-saas

@avilum One problem with your implementation that the service is that it spawns a new llama.cpp instance for every http request, so it can take a long time to respond (model has to be loaded every time).

I suggest you try using this branch with a different approach: Start llama.cpp once, and on every http request you create a TCP connection to the singleton instance. This will get you a clean environment for every request without the overhead of reloading the model.

**Comment by avideci** [comment]
> > If anyone is seeking for working client/server implementation, I wrote a minimal Go/Python server and client with live streaming, that is based on this awesome repo. See https://github.com/avilum/llama-saas
> 
> @avilum One problem with your implementation that the service spawns a new llama.cpp instance for every http request, so it can take a long time to respond (model has to be loaded every time).
> 
> I suggest you try using this branch with a different approach: Start llama.cpp once, and on every http request you create a TCP connection to the singleton instance. This will get you a clean environment for every request without the overhead of reloading the model.

For the sake of POC I ran the process every prompt.
I most definitrly Agree - I am working at the moment on loaded the DLL only once (main exe) and once this is merged, I might evensend the data over this TCP socket.
I'll load the model and then feed it with inputs, from the Go code, over any kind of socket / IPC.

**Comment by jart** [comment]
I disagree with this change. This is a large rearchitecting of the project that fundamentally changes its vision. No one will run your daemon. What value does having a TCP server mode offer, aside from fixing loading time? The issue of loading time is solved by #91 which we've implemented in the mmap branch: https://github.com/ggerganov/llama.cpp/tree/mmap It will be much easier to support win32 using mmap than it would be to support winsock.

**Comment by anzz1** [comment]
Why even the need of spawning multiple processes instead of multiple threads? Threads are already built-in, unlike sockets or mmap. The only truly portable thing is standard I/O, which can be redirected and also easily communicated with using simple file streams which are supported by everything. Instead of changing the main implementation much at all, you could just build any modules *outside* the main implementation as modules and communicate using these simple file streams. The main llama input would not need any sort of 'protocol' but just listen to '\n' or EOF like it currently does, and whatever modules could just follow that paradigm while communicating through the file streams. Am I missing something here?

**Comment by spirobel** [comment]
@jart 
> I disagree with this change. This is a large rearchitecting of the project that fundamentally changes its vision.

That is true.  It is better to keep the scope focused and make sure llama.cpp is as stable as possible.

@tarruda 
> While the protocol is a bit "low level", it should be easy to write a higher level API on top of this, such as a node.js web server or next.js app.

We can bundle llama with node.js in its current form. There is a library called caxa: https://github.com/leafac/caxa

It bundles any nodejs app into a single executable.  As a side effect of the way it is architected it unzips any executeables from the dist folder at runtime. 

So we can just place a compiled llama version in there and bundle it together. Then we can build a rest-api in node that can be easily called from nextjs.

I also found a way that turns an openapi.yml file into the single source of truth for the nodejs routes. 
The result of this is,
 that we get an interactive docs page for free that spits out curl commands to interact with the api. 

Here is the repo that combines caxa and this technique to make a single binary rpc daemon: https://github.com/spirobel/monerochan-merchant-rpc 

If there is an interest I can make something like this for llama.cpp.

If we make a one click template for a provider like digital ocean people could spin up their own on demand instances that work just like the openAI api.

**Comment by tarruda** [comment]
> This is a large rearchitecting of the project that fundamentally changes its vision

@jart @spirobel there's no rearchitecting here, in fact my goal was to introduce a server/client model with minimal changes. If you don´t pass the `-l` option, the app works exactly as it is now. This PR might seem like it is changing a lot, but if you review the commits individually, you will see there's not much change to existing logic.

The way this was implemented was basically by moving all the code (except for model loading code) from the `main` function to another reusable function called `llama_main`, then replacing the direct std streams references  by parameters passed to it. So the same code runs unmodified on `stdin/stdout/stderr` or on a TCP connection.

> What value does having a TCP server mode offer, aside from fixing loading time?

There are more unexplored applications of this TCP server mode. Here's a few ideas:

- Wrap into an http/json/rest/websocket server for using in a web application. Eventually I would like to write an API that is compatible with ChatGPT's API.
- Do any kind of processing (such as preload the prompt) prior to accepting connections/requests. This would allow you to essentially respond to queries more quickly by preloading a chatbot prompt, then for each request just process user input.
- Server/client usage, you can run llama.cpp in a more powerful computer and share with other devices in the same LAN.

@ggerganov also thinks this is a good idea: https://github.com/ggerganov/llama.cpp/pull/267#issuecomment-1474916599

I don't think this PR is mutually exclusive with the work you are doing, it is still useful to load the model faster on new process startups.

**Comment by spirobel** [comment]
@tarruda 
>  Do any kind of processing (such as preload the prompt) prior to accepting connections/requests. This would allow you to essentially respond to queries more quickly by preloading a chatbot prompt, then for each request just process user input.

Responsiveness and proper concurrency is always good! 😀👍 Nothing more frustrating than lagging or hung up programs.


That being said,

>Wrap into an http/json/rest/websocket server for using in a web application. Eventually I would like to write an API that is compatible with ChatGPT's API.

please take a look at the single binary rpc that I built! I am happy to answer any questions! I genuinely believe it is  better to implement this in nodejs instead of doing it in cpp. Monero also has a cpp rest rpc daemon and it is not fun to work with. It always hangs up or becomes unresponsive for long times. Also the documentation is always out of date. 
Using openapi (not openai 😅) as a single source of truth for routing and documentation solves this issue permanently.

> We can bundle llama with node.js in its current form. There is a library called caxa: https://github.com/leafac/caxa
> 
> It bundles any nodejs app into a single executable. As a side effect of the way it is architected it unzips any executeables from the dist folder at runtime.
> 
> So we can just place a compiled llama version in there and bundle it together. Then we can build a rest-api in node that can be easily called from nextjs.
> 
> I also found a way that turns an openapi.yml file into the single source of truth for the nodejs routes. The result of this is, that we get an interactive docs page for free that spits out curl commands to interact with the api.
> 
> Here is the repo that combines caxa and this technique to make a single binary rpc daemon: https://github.com/spirobel/monerochan-merchant-rpc
> 
> If there is an interest I can make something like this for llama.cpp.
> 
> If we make a one click template for a provider like digital ocean people could spin up their own on demand instances that work just like the openAI api.

**Comment by tarruda** [comment]
> implement this in nodejs instead of doing it in cpp


@spirobel  If you want to implement a server mode in another program/language such as node.js and without changes to llama.cpp, there's two ways I see you can go about it:

- Spawn a single instance of llama.cpp at server startup, and synchronize access to stdin/stdout between all requests. This might work well for a single user, but if llama.cpp the process crashes, you would have to restart it. Not to mention you would not be able to customize parameters per connection/request.
- Spawn a new llama.cpp instance for each request (this is what @avilum [did](https://github.com/ggerganov/llama.cpp/pull/278#issuecomment-1475390306)), which would be very slow with the way the model is loaded right now (also requires duplicating the model in memory). I'm not familiar with the `mmap` loading solution, but assuming it makes loading the model instant, would it be able to preload the prompt and/or customize the prompt/seed per request? Preloading the prompt is not implemented right now but it is something that can be done in the server mode I implemented here.

I agree that higher level abstractions are better done in platforms like node.js or python, but in this case I don't think it would be possible to implement a server in purely node.js and have the same efficiency of a fork/connection server approach.

Now, here is the last paragraph of the PR description

> While the protocol is a bit "low level", it should be easy to write a higher level API on top of this, such as a node.js web server or next.js app."

As you can see, my goal with this PR was to provide a base server/protocol that can be wrapped in a higher level API.  In fact I implemented the parameter passing protocol in a way to allow reusing the existing `gpt_params_parse` function, so any new parameters added are automatically supported. The server mode is basically a "zygote" for quickly spawning new instances of llama.cpp with custom parameters (except for `n_ctx`, which has to be passed to `ggml_init`). 

Technically I could have implemented a higher level abstraction such as a simple HTTP endpoint that parses json messages, but it would require me to add at least a JSON parsing library, which goes against the [goals](https://github.com/ggerganov/llama.cpp#description) of the project ("no dependencies"). I also think we would lose some flexibility by making assumptions about the format of the API, better to have a lower level implementation that can be tailored for various purposes.

Since this server mode is meant to be wrapped in a higher level abstraction, it might be better to implement it using Unix sockets instead of TCP sockets, which I might do later after getting some feedback. This is still an experiment/POC.

**Comment by anzz1** [comment]
@tarruda Please check this comment https://github.com/ggerganov/llama.cpp/issues/23#issuecomment-1477080912
If the TCP server was refactored as a module and use a generic C model of sharing the preloaded model inside the memory of a single process and use threads for multi-serve capability instead, foregoing the use of fork(), I could help porting this to winsock.

**Comment by j-f1** [comment]
I’m personally uncomfortable with this because I don’t believe new C code should be exposed directly to the internet, especially when we won’t be using security validated code to handle the network requests for us. I think this project should focus on offering a robust C-based API and then leave tasks like a HTTP API to inherently safer languages.

**Comment by KyL0N** [comment]
> I’m personally uncomfortable with this because I don’t believe new C code should be exposed directly to the internet, especially when we won’t be using security validated code to handle the network requests for us. I think this project should focus on offering a robust C-based API and then leave tasks like a HTTP API to inherently safer languages.

@j-f1 Yeah, one solution to address your concern is to create a C-based API that can be used internally within the system, while using a different language such as Python/Node.js to handle external network requests.

**Comment by tarruda** [comment]
> I’m personally uncomfortable with this because I don’t believe new C code should be exposed directly to the internet

This TCP mode is not meant to be used directly, in my previous comments I've hinted that I created this as a lower level protocol meant to be wrapped in a higher level solution, possibly written in Node.js or Python.

**Review by ggerganov — CHANGES_REQUESTED** [review]
Great job! Haven't had the chance to test it yet, but it seems well done.

I would like this to become a standalone example in the "./examples" folder.
The `main.cpp` example has to remain the way it is on `master`.
Even if you have to duplicate the code from `main` in the `tcp_server` example - that is OK.

**Comment by tarruda** [comment]
> I would like this to become a standalone example in the "./examples" folder.
The main.cpp example has to remain the way it is on master.
Even if you have to duplicate the code from main in the tcp_server example - that is OK.

@ggerganov I'm not sure if I understand. Do you want me to copy all the code in main.cpp to tcp_server.cpp and have it become a standalone program?

**Comment by tarruda** [comment]
@ggerganov  These are the changes I did to `main`:

  - Moved main I/O loop out of the `main()` function into a `run()` function which can be reused
  - In the `run()` function, changed `fprintf(stderr,...` to `fprintf(errstream, ...` where `errstream` is a `FILE *` parameter (`main()` passes `stderr`)
  - In the `run()` function, changed `printf(...` to `fprintf(outstream, ...` where `outstream` is a `FILE *` parameter (`main()` passes `stdout`)
  - In the `run()` function, changed `std::getline(std::cin,...` to `std::getline(instream,...` where `instream` is a `std::istream` parameter (`main()` passes `std:cin`)
  
> The main.cpp example has to remain the way it is on master.

Do you mean I should revert the changes listed above? These changes are mandatory for the implementation of the tcp server.

**Comment by ggerganov** [comment]
@tarruda 

> @ggerganov I'm not sure if I understand. Do you want me to copy all the code in main.cpp to tcp_server.cpp and have it become a standalone program?

Yes, the `main.cpp` is the example that everybody will run first when they clone the project. It has to be straightforward, demonstrating basic usage of `llama.cpp`. The `run()` abstraction is not necessary here.

**Comment by tarruda** [comment]
> The run() abstraction is not necessary here.

The run abstraction is necessary if we want to share the main loop with the tcp server, it is not practical for me to copy all the code in `main` and keep duplicating the changes back to a separate program whenever the main loop is updated. I will maintain this in my own private fork then.

**Comment by tkafka** [comment]
@ggerganov I feel that we are losing a lot here - I, for one, would love to be able to use @tarruda's fork with the API. Is there a way to add the API to this project?

**Comment by tarruda** [comment]
@tkafka I'm still maintaining these changes in my fork, and will keep rebasing for the foreseeable future (might even set up some script to do this semi-automatically in a daily basis). Here's the code: https://github.com/tarruda/llama.cpp/tree/tcp_server  just rebased it.

**Comment by sowa705** [comment]
Hi, i use your TCP fork and its working very well for my usecase This is a very important feature that should be merged imo

**Comment by jessejohnson** [comment]
Happy to see this as a separate llama.cpp-based project. It's nice to keep llama.cpp as a useful base to build on top of. See gpt4all for example. 🚀

**Pull request closed without merging**